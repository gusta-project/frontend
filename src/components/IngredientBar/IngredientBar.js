import PropTypes from 'prop-types';
import React, { Component, Fragment, createRef } from 'react';
import { ProgressBar, Tooltip, Overlay } from 'react-bootstrap';

export default class IngredientBar extends Component {
  static propTypes = {
    nicotine: PropTypes.number.isRequired,
    flavor: PropTypes.number.isRequired,
    vg: PropTypes.number.isRequired,
    pg: PropTypes.number.isRequired
  };

  static defaultProps = {
    nicotine: 0,
    flavor: 0,
    vg: 0,
    pg: 0
  };

  constructor(props) {
    super(props);

    this.refList = {
      nicotine: createRef(),
      flavor: createRef(),
      vg: createRef(),
      pg: createRef()
    };
    this.state = {
      nicotine: false,
      flavor: false,
      vg: false,
      pg: false
    };
    this.toggleTooltipVisibility = this.toggleTooltipVisibility.bind(this);
  }

  toggleTooltipVisibility(name) {
    this.setState({
      [name]: !this.state[name]
    });
  }

  tooltip(name) {
    const { [name]: show } = this.state;
    const { [name]: value } = this.props;
    const {
      [name]: { current: ref }
    } = this.refList;

    return (
      <Overlay target={ref} show={show} placement="top">
        <Tooltip id={`${name}-tooltip`}>{Math.round(value)}%</Tooltip>
      </Overlay>
    );
  }

  render() {
    const { nicotine, flavor, vg, pg } = this.props;

    const nicotineLabel =
      nicotine > 5 ? `Nic (${Math.round(nicotine)}%)` : 'Nic';
    const flavorLabel = flavor > 5 ? `Flavor (${Math.round(flavor)}%)` : 'Flv';
    const vgLabel = vg > 5 ? `VG (${Math.round(vg)}%)` : 'VG';
    const pgLabel = pg > 5 ? `PG (${Math.round(pg)}%)` : 'PG';

    return (
      <Fragment>
        <ProgressBar className="ingredient-bar">
          <ProgressBar
            now={nicotine}
            variant="danger"
            label={nicotineLabel}
            ref={this.refList.nicotine}
            onMouseEnter={() => this.toggleTooltipVisibility('nicotine')}
            onMouseLeave={() => this.toggleTooltipVisibility('nicotine')}
          />
          <ProgressBar
            now={flavor}
            label={flavorLabel}
            ref={this.refList.flavor}
            onMouseEnter={() => this.toggleTooltipVisibility('flavor')}
            onMouseLeave={() => this.toggleTooltipVisibility('flavor')}
          />
          <ProgressBar
            now={vg}
            variant="success"
            label={vgLabel}
            ref={this.refList.vg}
            onMouseEnter={() => this.toggleTooltipVisibility('vg')}
            onMouseLeave={() => this.toggleTooltipVisibility('vg')}
          />
          <ProgressBar
            now={pg}
            variant="warning"
            label={pgLabel}
            ref={this.refList.pg}
            onMouseEnter={() => this.toggleTooltipVisibility('pg')}
            onMouseLeave={() => this.toggleTooltipVisibility('pg')}
          />
        </ProgressBar>
        {this.tooltip('nicotine')}
        {this.tooltip('flavor')}
        {this.tooltip('vg')}
        {this.tooltip('pg')}
      </Fragment>
    );
  }
}
