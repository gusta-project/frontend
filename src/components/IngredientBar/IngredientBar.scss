.ingredient-bar {
  height: 32px;
  margin: 0 0 $base-spacing;
}
