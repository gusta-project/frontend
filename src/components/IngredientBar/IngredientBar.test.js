import React from 'react';
import renderer from 'react-test-renderer';

import IngredientBar from './IngredientBar';

describe('<IngredientBar />', () => {
  const props = {
    nicotine: 20,
    flavor: 10,
    vg: 60,
    pg: 10
  };

  it('renders correctly', () => {
    const component = renderer.create(<IngredientBar {...props} />);

    expect(component.toJSON()).toMatchSnapshot();
  });
});
