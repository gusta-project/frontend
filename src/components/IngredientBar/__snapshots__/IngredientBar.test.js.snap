// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IngredientBar /> renders correctly 1`] = `
<div
  className="ingredient-bar progress"
>
  <div
    aria-valuemax={100}
    aria-valuemin={0}
    aria-valuenow={20}
    className="progress-bar bg-danger"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    role="progressbar"
    style={
      Object {
        "width": "20%",
      }
    }
  >
    Nic (20%)
  </div>
  <div
    aria-valuemax={100}
    aria-valuemin={0}
    aria-valuenow={10}
    className="progress-bar"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    role="progressbar"
    style={
      Object {
        "width": "10%",
      }
    }
  >
    Flavor (10%)
  </div>
  <div
    aria-valuemax={100}
    aria-valuemin={0}
    aria-valuenow={60}
    className="progress-bar bg-success"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    role="progressbar"
    style={
      Object {
        "width": "60%",
      }
    }
  >
    VG (60%)
  </div>
  <div
    aria-valuemax={100}
    aria-valuemin={0}
    aria-valuenow={10}
    className="progress-bar bg-warning"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    role="progressbar"
    style={
      Object {
        "width": "10%",
      }
    }
  >
    PG (10%)
  </div>
</div>
`;
